src/runtime_src/core/pcie/linux/shim.cpp

```diff
@@ -211,53 +462,75 @@ class queue_cb {
             }
         }
 
+        if (rc < min_compl && qAioBatchEn) {
+             /* timeout happened, check if there is any io submission errors */
+             rc = check_io_submission_error(max_compl - num_evt, comps + num_evt);
+             num_evt += rc;
+        }
+
+        *actual = num_evt;
         return 0;
     }
 
     // submit the read/write i/o to the queue
     ssize_t queue_submit_io(xclQueueRequest *wr, aio_context_t *mAioCtx)
     {
         ssize_t rc = 0;
-        aio_context_t *aio_ctx = qAioEn ? &qAioCtx : mAioCtx;
+        bool aio = (wr->flag & XCL_QUEUE_REQ_NONBLOCKING) ? true : false;
 
-        for (unsigned int i = 0; i < wr->buf_num; i++) {
-            void *buf = (void *)wr->bufs[i].va;
-            struct iovec iov[2];
-            struct xocl_qdma_req_header header;
-
-            header.flags = wr->flag;
-            iov[0].iov_base = &header;
-            iov[0].iov_len = sizeof(header);
-            iov[1].iov_base = buf;
-            iov[1].iov_len = wr->bufs[i].len;
-
-            if (wr->flag & XCL_QUEUE_REQ_NONBLOCKING) {
-                struct iocb cb;
-                struct iocb *cbs[1];
-
-                memset(&cb, 0, sizeof(cb));
-                cb.aio_fildes = (int)qhndl;
-                cb.aio_lio_opcode = h2c ? IOCB_CMD_PWRITEV : IOCB_CMD_PREADV;
-                cb.aio_buf = (uint64_t)iov;
-                cb.aio_offset = 0;
-                cb.aio_nbytes = 2;
-                cb.aio_data = (uint64_t)wr->priv_data;
+        if (qAioBatchEn) {
+            /* queue up this async i/o request */
+            if (aio) {
+                return queue_up_request(wr);
+            } else {
+                /* this is synchronous i/o, flush all of the queued requests
+                 * first to maintain the order */
+                std::lock_guard<std::mutex> lk(reqLock);
+                while (!reqList.empty()) {
+   		    rc = queue_flush_aio_request();
+                    if (rc < 0)
+                        return rc;
+                }
+
+                /* fall through to process this request */
+                rc = 0;
+            }
+        }
 
+        /* synchronous i/o or no batching configured on the queue, submit the
+         * i/o right away */
+        struct xocl_qdma_req_header header;
+        header.flags = wr->flag;
+        if (aio) {
+            aio_context_t *aio_ctx = qAioEn ? &qAioCtx : mAioCtx;
+            for (unsigned int i = 0; i < wr->buf_num; i++) {
+                struct iovec iov[2];
+		struct iocb cb;
+		struct iocb *cbs[1];
```

Comment:
Do it this way:
```
struct iocb* cbs[1] = {nullptr};
prepare_io(cbs, ...);
```
no need for line 513.

Reply:
struct iocb* cbs[1] = {nullptr};
prepare_io(cbs, ...);
would not compile, if I changes to prepare_io(*cbs, ....) then it compile, but failed the later io_submit() call

so I did 
struct iocb cb;
struct iocb *cbs[1] = {&cb};